stream_pool.go

```diff
@@ -0,0 +1,55 @@
+package dht
+
+import (
+	"context"
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+type streamPool struct {
+	mu sync.Mutex
+	m  map[peer.ID]map[*stream]struct{}
+}
+
+func (sp *streamPool) get(ctx context.Context, p peer.ID) (*stream, bool) {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	for ps := range sp.m[p] {
+		sp.deleteLocked(ps, p)
+		if ps.err() != nil {
+			// Stream went bad and hasn't deleted itself yet.
+			continue
```

Comment:
shouldn't we delete it then?

Reply:
It does delete itself, but it's a race condition.